ui/src/app/utils/cluster-initializer.tsx

```diff
@@ -0,0 +1,301 @@
+  //
+  // The workspace namespace to initialize a cluster for.
+  workspaceNamespace: string;
+  // Callback which is called every time the cluster updates its status. This callback is *not*
+  // called when no cluster is found.
+  onStatusUpdate?: (ClusterStatus) => void;
+  // An optional abort signal which allows the caller to abort the initialization process, including
+  // cancelling any outstanding Ajax requests.
+  abortSignal?: AbortSignal;
+
+  // Override options. These options all have sensible defaults, but may be overridden for testing
+  // or special scenarios (such as an initialization flow which should not take any actions).
+  initialPollingDelay?: number;
+  maxPollingDelay?: number;
+  overallTimeout?: number;
+  maxCreateCount?: number;
+  maxDeleteCount?: number;
+  maxResumeCount?: number;
+  maxServerErrorCount?: number;
+}
+
+/**
+ * A controller class implementing client-side logic to initialize a Leonardo cluster. This class
+ * will continue to poll the getCluster endpoint, taking certain actions as required to nudge the
+ * cluster towards a running state, and will eventually resolve the Promise with a running cluster,
+ * or otherwise reject it with information about the failure mode.
+ *
+ * This is an unusually heavyweight controller class on the client side. It's worth noting a couple
+ * reasons why we ended up with this design:
+ *  - Cluster initialization can take up to 10 minutes, which is beyond the scope of a single
+ *    App Engine server-side request timeout. To reliably implement this control on the server side
+ *    would likely require new database persistence, tasks queues, and additional APIs in order to
+ *    provide the client with status updates.
+ *  - Ultimately, we might expect this type of functionality ("Get me a cluster for workspace X and
+ *    bring it to a running state") to exist as part of the Leonardo application. So rather than
+ *    build our own server-side equivalent, we adopted a client-side solution as a holdover.
+ */
+export class ClusterInitializer {
+  // Core properties for interacting with the caller and the cluster APIs.
+  private workspaceNamespace: string;
+  private onStatusUpdate: (ClusterStatus) => void;
+  private abortSignal?: AbortSignal;
+
+  // Properties to track & control the polling loop. We use a capped exponential backoff strategy
+  // and a series of "maxFoo" limits to ensure the initialization flow doesn't get out of control.
+  private currentDelay: number;
+  private maxDelay: number;
+  private overallTimeout: number;
+  private maxCreateCount: number;
+  private maxDeleteCount: number;
+  private maxResumeCount: number;
+  private maxServerErrorCount: number;
+
+  // Properties to track progress, actions taken, and errors encountered.
+  private createCount = 0;
+  private deleteCount = 0;
+  private resumeCount = 0;
+  private serverErrorCount = 0;
+  private initializeStartTime?: number;
```

Comment:
Per my other comment - some of these properties could be made final or eliminated if we just had a single entrypoint for initializing the cluster (rather than the two phase create object-> run once)